### ISSUE TITLE
(feature request) dynamic patching - send to $activecanvas

### ISSUE THREAD

**Issue by ludnny** [pure-data/pure-data#1867]
Hi, 

It's possible via dynamic patching to create objects dynamically in chosen patches or sub-patches, using the pd-filename or a namecanvas.
But most of the time, we don't know beforehand in which patch/sub-patch we will need to use dynamic patching.

A variable like _$activecanvas_ would be quite handy to be able to send instructions to the active canvas, without having to use namecanvas.

```
;
pd-$activecanvas editmode1;
```

![image](https://user-images.githubusercontent.com/46849877/212985436-f7e072ed-c7fe-400d-95a4-a99ddeb19473.png)

Use-cases:
- toolbox to dynamically create objects in the active patch
- could help to build *intelligent-patching* tools inside Vanilla without the help of externals or tcl plugins.

It looks like dynamic patching is not considered as an official feature of PD, more like a hack. 
But it allows to go past a lot of limitations of PD, especially when externals are not an option.  
I hope these features will continue to be still developed.

**Comment by porres** [comment]
I think this PR should be useful and close this issue, what do you say? --> https://github.com/pure-data/pure-data/pull/837

**Comment by ludnny** [comment]
If I understand correctly, this PR adds an argument to the namecanvas object. 
Ideally, we would not need to use a namecanvas at all. 

An example:
- I have *mymainpatch.pd*, empty except for the subpatch *mytoolbox* (containing the dynamic patching stuff)
- I create another subpatch *mysubpatch*, and I start patching there. 
- I want to create a dynamic object at mouse position in this subpatch, so I press a keyboard shortcut (to call a function from *mytoolbox*, like `; $activecanvas obj mouseposX mouseposY myrandomobject`)

Would your PR work in this kind of case? 
Without having to create a namecanvas in *mysubpatch* beforehand?

**Comment by porres** [comment]
what external are you currently using now for this? if it's [sendcanvas], this PR is an alternative for it

**Comment by ludnny** [comment]
I just tried [iemguts/sendcanvas]. 
I looks super useful and would be an important addition to Vanilla.

But I don't see how that could solve my use case?
In my example, we don't know the depth or the relationship level between the active patch and the dynamic patching subpatch.

I'd like to be able to patch in any subpatch or in any new patch, and to be able to call the dynamic patching subpatch which can be anywhere else.

My example here for more clarity:
![image](https://user-images.githubusercontent.com/46849877/213019323-38ce1aa6-a0eb-48b0-9c07-dbe8e2d34222.png)

[activecanvas.pd.zip](https://github.com/pure-data/pure-data/files/10439642/activecanvas.pd.zip)

In term of UX, I don't see that solved by a new object or argument, but by a variable like $activecanvas 

![image](https://user-images.githubusercontent.com/46849877/213020684-a5687396-577b-4f8e-b7a0-32c2114f1387.png)

But maybe I misunderstood the use of [iemguts/sendcanvas] or the use of your PR?
Do you think that could work to solve this example?

**Comment by ludnny** [comment]
Another use-case for this request is Automatonism.
https://www.automatonism.com/
Automatonism is a great way to make PD more accessible, especially thanks to it's module browser. 

But the dynamic patching process behind this module browser make it very limited: the patch should be called main.pd and can't be renamed, impossibility to create modules in subpatches or in other patches, and the module is created at a predefined position (200x200) instead of mouse position. 

![image](https://user-images.githubusercontent.com/46849877/213498297-d8dee907-2c19-4534-bb51-67874734bc3a.png)

**Comment by porres** [comment]
> Do you think that could work to solve this example?

not really, sorry, I didn't pay close attention and thought you said you used an external for this and I thought that was it... so you currently do not rely on any external for this, right?

Seems like a special case and something that should probably need an external. I think of something like [cyclone/active] or [else/canvas.active] but that could give you the window name as well (as given by [else/canvas.name]. With this name you can use [send] to send messages to it for dynamic patching. I can try and see if I can make [canvas.active] output the name.

**Comment by porres** [comment]
Ok, added an outlet to canvas.active to output window name, you still need to trim the ".c" from it so you can use it as a send symbol, but it works. Should I include this to ELSE, are you happy with it? 

Again, seems overkill for a vanilla feature, it's not something that comes up really often.

![Screen Shot 2023-01-19 at 15 07 22](https://user-images.githubusercontent.com/5360142/213525512-cbc7dd04-6b9c-4c93-98fa-cbf6bf426c39.png)

**Comment by ludnny** [comment]
> Ok, added an outlet to canvas.active to output window name, you still need to trim the ".c" from it so you can use it as a send symbol, but it works. Should I include this to ELSE, are you happy with it?
> 
It looks absolutely perfect, thank you @porres !

> Again, seems overkill for a vanilla feature

I'd like to insist ;)

The beauty of Automatonism is how easy it is to start with. 
Install PureData, open the Automatonism patch, and play!
No dependency, no external. 
I'm pretty sure this kind of project and the ease of use and installation brought people to PD, making it more accessible.

A feature like this would remove the main limitation of Automatonism: freedom to patch everywhere, not having to stay on the same window, not having to give a special name to your patch. 

This is not a feature coming out of nowhere, it completes an already existing set of features offered by dynamic patching, making them more "dynamic".

> it's not something that comes up really often.

I'll continue to post some use-cases solved by this feature, I faced plenty of them in the past, and I saw some of them on forums!

In any case, thank you very much for canvas.active and this second outlet!

**Comment by ludnny** [comment]
> You can keep this active, it's not up to me, but if this stays open forever or gets officially rejected, would you be rather satisfied with what I'm offering now with [canvas.active]? Can I go on and officially add this functionality into it? It would be only for you...

I'm very satisfied by your solution with [canvas.active] in ELSE :)
Thank you very much, that's super impressive. 

I get your point as well about keeping Vanilla small and simple.

I'd like to post a few other use cases in this Issue, to document it a bit more, and to see if other users are lacking this feature in Vanilla as well or not.

**Comment by porres** [comment]
done with canvas.active, the -name flag now ouptus the canvas name - uploading commit soon!

![Screen Shot 2023-01-20 at 01 46 37](https://user-images.githubusercontent.com/5360142/213619288-b83f0d3f-2104-448d-a609-72ba11fd2ff8.png)

**Comment by ludnny** [comment]
> done with canvas.active, the -name flag now ouptus the canvas name - uploading commit soon!
> 
Perfect, thank you very much!


I add here another use case:
_A toolbar abstraction that allows you to automatically create objects, number boxes, messages, etc. in your patch._
![image](https://user-images.githubusercontent.com/46849877/213698596-4c82798c-c22d-448a-8d8c-4d043603122d.png)
https://www.youtube.com/watch?v=qIxhiCP242A

It was from another issue on this repository:
https://github.com/pure-data/pure-data/issues/794

The issue was closed because the proposal was to add a feature to [pdcontrol], but the use case itself has not been solved, and could be solved by [canvas.active].

**Comment by porres** [comment]
> The issue was closed because the proposal was to add a feature to [pdcontrol]

actually the solution was to add a 2nd argument to [namecanvas], so you can also just use [else/canvas.setname] instead of [else/canvas.active]. And I already have a PR to add this functionality to [namecanvas]/

**Comment by porres** [comment]
ELSE rc7 is out with this feature, can we just close this?

**Comment by porres** [comment]
could we really close this @ludnny or anyone else?

**Comment by ludnny** [comment]
> could we really close this @ludnny or anyone else?

This is still a feature I'd need in Vanilla. 

That's super helpful to have an alternative using an external, thanks again for ELSE/canvas.active @porres , but as mentionned in the first post the goal of the request was:

> to build intelligent-patching tools inside Vanilla without the help of externals or tcl plugins

**Comment by porres** [comment]
I am asking you to be realistic and realise that it's not gonna happen :)
